README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.
For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).
For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

# Multiple Pile Faces

**Multiple Pile Faces** package lets you display a stacked list of users/persons in your Flutter app.

## Features

The `FacePile` widget is built to be `a unique stacked list` that can be used in your Flutter app.This widget is often used to assign tasks to individuals inside a workflow or to share who has access to a certain view or file. Each circle has a picture of the person it symbolizes.

The package will handle the animation by itself.

![thumb](https://user-images.githubusercontent.com/68671238/193428931-13acc0a5-a9dd-45b8-ab18-db9b1111e0f4.png)

<hr>

## Getting started

1. Add the latest version of package to your `pubspec.yaml` (and run `dart pub get`):

```dart
dependencies:
  multi_pile_faces: ^1.0.0
```

2. Import the package and use it in your Flutter App.

```dart
import 'package:multi_pile_faces/multi_pile_faces.dart';
```

<hr>

## Usage

There are a number of properties that you can modify:

- faces (List of users)
- faceSize (Avatar size)
- facePercentOverlap (Avatar percent overlap)
- animationDuration (Animation duration)
- ... etc.

**Example Usage ( complete with all params ):**

<table>
 <tr>
 <td>

```dart

  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: Center(
        child: ConstrainedBox(
          constraints: const BoxConstraints(maxWidth: 200),
          child: FacePile(
            faces: _allUsers,
            faceSize: 42,
            facePercentOverlap: .4,
            borderColor: Colors.white,
            nameLabelColor: Colors.black,
            borderWidth: 0.4,
            animationDuration: const Duration(milliseconds: 500),
          ),
        ),
      ),
      floatingActionButton: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          Padding(
            padding: const EdgeInsets.only(right: 18.0),
            child: ElevatedButton(
              child: const Text('Remove User Icon'),
              style: ElevatedButton.styleFrom(
                primary: Colors.red,
              ),
              onPressed: _removeUserFromPile,
            ),
          ),
          FloatingActionButton(
            onPressed: _addUserToPile,
            mini: true,
            backgroundColor: Colors.white,
            child: const Icon(Icons.add, color: Colors.black,),
          ),
        ],
      ),
    );
  }

```
   </td>
   <td>
     Here's what it looks like:

<hr>




https://user-images.githubusercontent.com/68671238/193429465-5f83de54-d103-45d9-a3da-78937372cd63.mp4






   </td>
  </tr>
  </table>
<hr>

## Next Goals
We are working on some improvements including:

- [ ] Make the widget more customizable.

## Issues & Feedback
Please file an [issue!](https://github.com/aliMissaoui/Flutter---Multiple-Pile-Faces-Package/issues) to send feedback or report a bug. Thank you!

```


CHANGELOG.md
## 1.0.0

* Multiple pile faces.


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:multi_pile_faces/multi_pile_faces.dart';
import 'home_page.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      debugShowCheckedModeBanner: false,
      home: const MyHomePage(),
    );
  }
}



lib/src/face_pile_widget.dart
import 'package:flutter/material.dart';
import 'appearing_and_disappearing_face_widget.dart';
import 'face_holder.dart';

class FacePile extends StatefulWidget {
  /// List of user profile image.
  final List<FaceHolder> faces;

  /// Avatar size [default -> 48]
  final double faceSize;

  /// Avatar percent overlap [default -> 0.1 or (10%)]
  final double facePercentOverlap;

  /// Color of the name label
  final Color nameLabelColor;

  /// Background color of the avatar [default -> #222222]
  final Color backgroundColor;

  /// Name label font size [default -> 8.0]
  final double nameFontSize;

  /// Name label font width [default -> bold]
  final FontWeight nameFontWeight;

  /// Avatar border width [default -> 0.0]
  final double borderWidth;

  /// Avatar border color [default -> white]
  final Color borderColor;

  /// Animation duration [default -> 250 microseconds]
  final Duration animationDuration;

  ///A group of overlapping rounding avatars are called face piles,
  ///[FacePile] displays a stacked list of avatars,
  ///Each circle represents a person and contains their image.
  const FacePile({
    Key? key,
    required this.faces,
    this.facePercentOverlap = .1,
    this.faceSize = 48,
    this.nameLabelColor = const Color(0xFF222222),
    this.backgroundColor = const Color(0xFF888888),
    this.nameFontSize = 8.0,
    this.nameFontWeight = FontWeight.bold,
    this.borderColor = Colors.white,
    this.borderWidth = 0.0,
    this.animationDuration = const Duration(milliseconds: 250),
  }) : super(key: key);

  @override
  State<FacePile> createState() => _FacePileState();
}

class _FacePileState extends State<FacePile> {
  final _visibleFaces = <FaceHolder>[];

  void _syncFacesWithPile() {
    setState(() {
      final newFaces = widget.faces.where(
        (face) =>
            _visibleFaces.where((visibleFace) => visibleFace == face).isEmpty,
      );

      for (final newFace in newFaces) {
        _visibleFaces.add(newFace);
      }
    });
  }

  @override
  void initState() {
    super.initState();

    WidgetsBinding.instance?.addPostFrameCallback((timeStamp) {
      _syncFacesWithPile();
    });
  }

  @override
  void didUpdateWidget(covariant FacePile oldWidget) {
    super.didUpdateWidget(oldWidget);

    WidgetsBinding.instance?.addPostFrameCallback((timeStamp) {
      _syncFacesWithPile();
    });
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        final facesCount = _visibleFaces.length;

        double facePercentVisible = 1.0 - widget.facePercentOverlap;
        final maxIntrinsicWidth = facesCount > 1
            ? (1 + (facePercentVisible * (facesCount - 1))) * widget.faceSize
            : widget.faceSize;

        late double leftOffset;
        if (maxIntrinsicWidth > constraints.maxWidth) {
          leftOffset = 0;
          facePercentVisible =
              ((constraints.maxWidth / widget.faceSize) - 1) / (facesCount - 1);
        } else {
          leftOffset = (constraints.maxWidth - maxIntrinsicWidth) / 2;
        }

        if (constraints.maxWidth < widget.faceSize) {
          return const SizedBox();
        }

        return SizedBox(
          height: widget.faceSize,
          child: Stack(
            clipBehavior: Clip.none,
            children: [
              for (var i = 0; i < facesCount; i += 1)
                AnimatedPositioned(
                  key: ValueKey(_visibleFaces[i].id), // or use an id!
                  top: 0,
                  left: leftOffset + (i * facePercentVisible * widget.faceSize),
                  width: widget.faceSize,
                  height: widget.faceSize,
                  duration: widget.animationDuration,
                  curve: Curves.easeInOut,
                  child: AppearingAndDisappearingFace(
                    face: _visibleFaces[i],
                    faceSize: widget.faceSize,
                    showFace: widget.faces.contains(_visibleFaces[i]),
                    backgroundColor: widget.backgroundColor,
                    nameLabelColor: widget.nameLabelColor,
                    onDisappear: () {
                      setState(() {
                        _visibleFaces.removeAt(i);
                      });
                    },
                    borderColor: widget.borderColor,
                    borderWidth: widget.borderWidth,
                    animationDuration: widget.animationDuration,
                  ),
                )
            ],
          ),
        );
      },
    );
  }
}


lib/src/avatar_circle_widget.dart
import 'package:flutter/material.dart';
import 'package:transparent_image/transparent_image.dart';
import 'face_holder.dart';

class AvatarCircle extends StatelessWidget {
  final FaceHolder faceHolder;
  final double size;
  final Color nameLabelColor;
  final Color backgroundColor;
  final Color borderColor;
  final double borderWidth;
  final Duration animationDuration;

  const AvatarCircle({
    Key? key,
    required this.faceHolder,
    required this.size,
    required this.nameLabelColor,
    required this.backgroundColor,
    required this.borderColor,
    required this.borderWidth,
    required this.animationDuration,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      width: size,
      height: size,
      decoration: BoxDecoration(
        shape: BoxShape.circle,
        border: Border.all(
          width: borderWidth,
          color: borderColor,
        ),
        color: backgroundColor,
        boxShadow: [
          BoxShadow(
            color: Colors.black.withOpacity(0.2),
            blurRadius: 5,
            offset: const Offset(0, 3),
          )
        ],
      ),
      clipBehavior: Clip.none,
      child: Stack(
        children: [
          Center(
            child: Padding(
              padding: const EdgeInsets.symmetric(horizontal: 8.0),
              child: Text(
                faceHolder.name,
                maxLines: 1,
                overflow: TextOverflow.ellipsis,
                style: TextStyle(
                  color: nameLabelColor,
                  fontSize: 8,
                  fontWeight: FontWeight.bold,
                ),
              ),
            ),
          ),
          ClipRRect(
            borderRadius: BorderRadius.circular(50),
            child: FadeInImage(
              width: size,
              height: size,
              placeholder: MemoryImage(kTransparentImage),
              image: faceHolder.avatar,
              fit: BoxFit.contain,
              fadeInDuration: animationDuration,
            ),
          )
        ],
      ),
    );
  }
}


lib/src/appearing_and_disappearing_face_widget.dart
import 'package:flutter/material.dart';
import 'avatar_circle_widget.dart';
import 'face_holder.dart';

class AppearingAndDisappearingFace extends StatefulWidget {
  final FaceHolder face;
  final double faceSize;
  final bool showFace;
  final Color nameLabelColor;
  final Color backgroundColor;
  final VoidCallback onDisappear;
  final Color borderColor;
  final double borderWidth;
  final Duration animationDuration;

  const AppearingAndDisappearingFace({
    Key? key,
    required this.face,
    required this.faceSize,
    required this.showFace,
    required this.onDisappear,
    required this.nameLabelColor,
    required this.backgroundColor,
    required this.borderColor,
    required this.borderWidth,
    required this.animationDuration,
  }) : super(key: key);

  @override
  State<AppearingAndDisappearingFace> createState() =>
      _AppearingAndDisappearingFaceState();
}

class _AppearingAndDisappearingFaceState
    extends State<AppearingAndDisappearingFace>
    with SingleTickerProviderStateMixin {
  void _syncScaleAnimationWithWidget() {
    if (widget.showFace &&
        !_scaleController.isCompleted &&
        _scaleController.status != AnimationStatus.forward) {
      _scaleController.forward();
    } else if (!widget.showFace &&
        !_scaleController.isDismissed &&
        _scaleController.status != AnimationStatus.reverse) {
      _scaleController.reverse();
    }
  }

  late AnimationController _scaleController;
  late Animation _scaleAnimation;

  @override
  void initState() {
    super.initState();

    _scaleController = AnimationController(
        vsync: this, duration: const Duration(milliseconds: 500))
      ..addStatusListener((status) {
        if (status == AnimationStatus.dismissed) {
          widget.onDisappear();
        }
      });
    _scaleAnimation = CurvedAnimation(
      parent: _scaleController,
      curve: Curves.elasticOut,
    );

    _syncScaleAnimationWithWidget();
  }

  @override
  void didUpdateWidget(AppearingAndDisappearingFace oldWidget) {
    super.didUpdateWidget(oldWidget);

    _syncScaleAnimationWithWidget();
  }

  @override
  void dispose() {
    _scaleController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: widget.faceSize,
      height: widget.faceSize,
      child: Center(
        child: AnimatedBuilder(
          animation: _scaleAnimation,
          builder: (context, child) {
            return Transform.scale(
              scale: _scaleAnimation.value,
              child: AvatarCircle(
                faceHolder: widget.face,
                size: widget.faceSize,
                nameLabelColor: widget.nameLabelColor,
                backgroundColor: widget.backgroundColor,
                borderColor: widget.borderColor,
                borderWidth: widget.borderWidth,
                animationDuration: widget.animationDuration,
              ),
            );
          },
        ),
      ),
    );
  }
}


example/lib/home_page.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:multi_pile_faces/multi_pile_faces.dart';

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  late List<FaceHolder> _availableUsers;
  final _allUsers = <FaceHolder>[
    FaceHolder(
    id: '0',
    name: 'first user',
    avatar: const AssetImage('assets/images/avatar.jpg'),
  ),
    FaceHolder(
      id: '1',
      name: 'second user',
      avatar: const AssetImage('assets/images/avatar1.png'),
    ),
    FaceHolder(
      id: '2',
      name: 'third user',
      avatar: const AssetImage('assets/images/avatar2.png'),
    ),
    FaceHolder(
      id: '3',
      name: 'fourth user',
      avatar: const AssetImage('assets/images/avatar3.jpg'),
    ),

  ];
  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    _availableUsers = List.from(_allUsers);
  }

  void _addUserToPile() {
    if (_availableUsers.isNotEmpty) {
      final user = _availableUsers.removeLast();
      setState(() {
        _allUsers.add(user);
      });
    }
  }

  void _removeUserFromPile() {
    if (_allUsers.isNotEmpty) {
      setState(() {
        final user = _allUsers.removeAt(_allUsers.length-1);
        _availableUsers.insert(0, user);
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: Center(
        child: ConstrainedBox(
          constraints: const BoxConstraints(maxWidth: 200),
          child: FacePile(
            faces: _allUsers,
            faceSize: 42,
            facePercentOverlap: .4, // 40% face overlap.
            borderColor: Colors.white,
            nameLabelColor: Colors.black,
            borderWidth: 0.4,
            animationDuration: const Duration(milliseconds: 500),
          ),
        ),
      ),
      floatingActionButton: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          Padding(
            padding: const EdgeInsets.only(right: 18.0),
            child: ElevatedButton(
              child: const Text('Remove User Icon'),
              style: ElevatedButton.styleFrom(
                primary: Colors.red,
              ),
              onPressed: _removeUserFromPile,
            ),
          ),
          FloatingActionButton(
            onPressed: _addUserToPile,
            mini: true,
            backgroundColor: Colors.white,
            child: const Icon(Icons.add, color: Colors.black,),
          ),
        ],
      ),
    );
  }
}
